gtk/tests_backend/dialogs.py

```diff
@@ -21,23 +22,42 @@ def _default_close_handler(self, dialog, gtk_result):
         dialog._impl.native.response(gtk_result)
         self._wait_for_dialog("Wait for dialog to disappear")
 
-    def _setup_dialog_result(self, dialog, gtk_result, close_handler=None):
+    def _setup_dialog_result(
+        self, dialog, gtk_result, close_handler=None, pre_close_test_method=None
+    ):
         # Install an overridden show method that invokes the original,
         # but then closes the open dialog.
         orig_show = dialog._impl.show
 
         def automated_show(host_window, future):
             orig_show(host_window, future)
 
-            if close_handler:
-                close_handler(dialog, gtk_result)
-            else:
-                self._default_close_handler(dialog, gtk_result)
+            async def _close_dialog():
+                # Add a slight delay for the dialog to show up
+                await asyncio.sleep(0.05)
+                try:
+                    if pre_close_test_method:
+                        pre_close_test_method()
+                finally:
+                    # Attempt to close the dialog regardless of any previous exceptions
+                    try:
+                        if close_handler:
+                            close_handler(dialog, gtk_result)
+                        else:
+                            self._default_close_handler(dialog, gtk_result)
+                    except Exception as e:
+                        # An error occurred closing the dialog; that means the dialog
+                        # isn't what as expected, so record that in the future.
+                        future.set_exception(e)
+
+            asyncio.create_task(_close_dialog(), name="close-dialog")
```

Comment:
Why has this become an async task?

Reply:
I had to add a slight delay to allow the dialog to show up:
https://github.com/beeware/toga/blob/f4d6c64766169cbc292701f9701997eb07b9063b/gtk/tests_backend/dialogs.py#L35-L53
So, I had made it an async task.